src/cds/src/main/kotlin/org/icpclive/cds/cats/CATSDataSource.kt

```diff
@@ -0,0 +1,175 @@
+package org.icpclive.cds.cats
+
+import kotlinx.coroutines.*
+import kotlinx.datetime.Clock
+import kotlinx.serialization.Serializable
+import org.icpclive.api.*
+import org.icpclive.cds.ContestParseResult
+import org.icpclive.cds.FullReloadContestDataSource
+import org.icpclive.cds.common.jsonLoaderService
+import org.icpclive.util.getCredentials
+import java.awt.Color
+import java.util.*
+import kotlin.time.Duration.Companion.hours
+import kotlin.time.Duration.Companion.seconds
+
+
+class CATSDataSource(val properties: Properties, creds: Map<String, String>) : FullReloadContestDataSource(5.seconds) {
+    private val login = creds["cats.login"]
+    private val password = creds["cats.password"]
+
+    // :TODO I may send less count of quires to the server but I need to know when my sid is expired
+    private val sid: String
+        get() = runBlocking {
+            async { authLoader.loadOnce().sid }.await()
+        }
+
+    @Serializable
+    data class Auth(val status: String, val sid: String, val cid: Long)
+
+    @Serializable
+    data class Problem(val id: Int, val name: String, val code: String)
+
+    @Serializable
+    data class Problems(val problems: ArrayList<Problem>)
+
+    @Serializable
+    data class Team(val id: Int, val account_id: Int, val login: String, val name: String, val role: String)
+
+    @Serializable
+    data class Users(val users: ArrayList<Team>)
+
+    @Serializable
+    data class Contest(
+        val title: String,
+        val start_date: String,
+        val freeze_date: String,
+        val finish_date: String,
+        val rules: String
+    )
+
+    @Serializable
+    data class Run(
+        val type: String,
+        val id: Int? = null,
+        val state_text: String? = null,
+        val problem_id: Int? = null,
+        val team_id: Int? = null,
+        val time: String? = null
+        // ::TODO add failed tests
+    )
+
+    private val authLoader =
+        jsonLoaderService<Auth> { "${properties.getProperty("url")}?f=login;login=$login;passwd=$password;json=1" }
+    private val problemsLoader =
+        jsonLoaderService<Problems> { "${properties.getProperty("url")}problems?cid=${properties.getProperty("cid")};sid=$sid;json=1" }
+    private val usersLoader =
+        jsonLoaderService<Users> { "${properties.getProperty("url")}users?cid=${properties.getProperty("cid")};sid=$sid;rows=1000;json=1" }
+    private val contestLoader =
+        jsonLoaderService<Contest> { "${properties.getProperty("url")}contest_params?cid=${properties.getProperty("cid")};sid=$sid;json=1" }
+    private val runsLoader =
+        jsonLoaderService<ArrayList<Run>> { "${properties.getProperty("url")}console?cid=${properties.getProperty("cid")};sid=$sid;rows=1000;json=1" }
+
+    override suspend fun loadOnce(): ContestParseResult {
+//        val element = problemsLoader.loadOnce()
```

Comment:
Please, don't leave commented code.

Reply:
Then what I need to do If I want to left some code which may be useful.

Reply:
Well. It depends. For example, here, you can load all 4 things to variables, not directly to function arguments. And add conditional logging in uncommented code.